jigglypuffRL/deeprl/vpg/vpg.py

```diff
@@ -0,0 +1,212 @@
+
+        # create distribution based on policy_fn output
+        action, c = self.policy_fn.sample_action(Variable(state))
+        val = self.value_fn(Variable(state))
+
+        # store policy probs and value function for current traj
+        self.policy_fn.policy_hist = torch.cat(
+            [
+                self.policy_fn.policy_hist,
+                c.log_prob(action).unsqueeze(0),
+            ]
+        )
+
+        self.value_fn.value_hist = torch.cat([self.value_fn.value_hist, val])
+
+        return action
+
+    # get clipped loss for single trajectory (episode)
+    def get_traj_loss(self):
+        R = 0
+        returns = []
+
+        # calculate discounted return
+        for r in self.policy_fn.traj_reward[::-1]:
+            R = r + self.gamma * R
+            returns.insert(0, R)
+
+        # advantage estimation
+        returns = torch.FloatTensor(returns).to(self.device)
+        A = Variable(returns) - Variable(self.value_fn.value_hist)
+
+        # compute policy and value loss
+        loss_policy = torch.sum(
+  	    torch.mul(self.policy_fn.policy_hist, A)
+	).mul(-1).unsqueeze(0)
+
+        loss_value = nn.MSELoss()(
+            self.value_fn.value_hist, Variable(returns)
+        ).unsqueeze(0)
+
+        # store traj loss values in epoch loss tensors
+        self.policy_fn.loss_hist = torch.cat([self.policy_fn.loss_hist, loss_policy])
+        self.value_fn.loss_hist = torch.cat([self.value_fn.loss_hist, loss_value])
+
+        # clear traj history
+        self.policy_fn.traj_reward = []
+        self.policy_fn.policy_hist = Variable(torch.Tensor())
+        self.value_fn.value_hist = Variable(torch.Tensor())
+
+    def update_policy(self, ep):
+        # mean of all traj losses in single epoch
+        loss_policy = torch.mean(self.policy_fn.loss_hist)
+        loss_value = torch.mean(self.value_fn.loss_hist)
+
+        # tensorboard book-keeping
+        if self.tensorboard_log:
+            self.writer.add_scalar("loss/policy", loss_policy, ep)
+            self.writer.add_scalar("loss/value", loss_value, ep)
+
+        # take gradient step
+        self.optimizer_policy.zero_grad()
+        loss_policy.backward()
+        self.optimizer_policy.step()
+
+        self.optimizer_value.zero_grad()
+        loss_value.backward()
+        self.optimizer_value.step()
+
+        # clear loss history for epoch
+        self.policy_fn.loss_hist = Variable(torch.Tensor())
+        self.value_fn.loss_hist = Variable(torch.Tensor())
+
+    def learn(self):
+        # training loop
+        for ep in range(self.epochs):
+            epoch_reward = 0
+            for i in range(self.actor_batch_size):
+                s = self.env.reset()
+                done = False
+                for t in range(self.timesteps_per_actorbatch):
+                    a = self.select_action(s)
+                    s, r, done, _ = self.env.step(np.array(a))
+
+                    if self.render:
+                        env.render()
```

Comment:
Should be `self.env()`

Reply:
aah sorry for this one will modify this one